### DISCUSSION TITLE
Using the MRTK for Touch Input

### DISCUSSION THREAD

**Discussion by gisornator** [Rufus31415/Simple-WebXR-Unity#46] [Q&A] [answered]
Hey,

I didn't want to crash the Hit-Test Issue with that question, since we got that covered for Android and it is working now. However, I would like to have some description on how to properly use the MRTK as input manager to interact with Unity UI elements during an XRSession.

I thought, that I had the MRTK set up properly but I'm recieving a warning now, which suggests, that something is not configured correctly in my project:

`[MRTK Issue] Touch 0 has already been registered with the Input Manager!`

![grafik](https://user-images.githubusercontent.com/17878740/150807663-a0b83c5b-98db-4a08-9d81-36d4d26d67df.png)

This is my MRTK config GameObject:
![grafik](https://user-images.githubusercontent.com/17878740/150807943-36d9e6ce-268e-439d-bc51-b4edc994ce09.png)

**Comment by Rufus31415** [comment]
Input sources are set up in "Input Data Provider" expandable menu that we see on your screenshot.
You should add the one provided with SimpleWebXR. You could also disable "Unity Touch Device Manager" It will probably remove your warning, but I don't understand why you get it without seeing the whole project

**Comment by gisornator** [comment]
We're getting closer I think... I removed the Unity Touch Device Manager and it removed the warning. But I noticed that I already had your Input provider set.
![grafik](https://user-images.githubusercontent.com/17878740/151312290-0b9f1c23-85de-4b77-b635-bbbf96be32e8.png)

**Comment by Rufus31415** [comment]
So it should work. A touch on the screen in immersive session should raise a select event :https://github.com/Rufus31415/Simple-WebXR-Unity/blob/a9060413037d6b5dd985abba29ed17fcd51ff79d/com.rufus31415.simplewebxr/Runtime/Plugins/WebGL/SimpleWebXR.jslib#L603

Then the ```_onSessionEvent``` handler is called and writes ```_controllerEvents``` array : https://github.com/Rufus31415/Simple-WebXR-Unity/blob/a9060413037d6b5dd985abba29ed17fcd51ff79d/com.rufus31415.simplewebxr/Runtime/Plugins/WebGL/SimpleWebXR.jslib#L507

At next frame, the event is propagated to ```_byteArray``` :https://github.com/Rufus31415/Simple-WebXR-Unity/blob/a9060413037d6b5dd985abba29ed17fcd51ff79d/com.rufus31415.simplewebxr/Runtime/Plugins/WebGL/SimpleWebXR.jslib#L419

```_byteArray``` is shared with C# code : https://github.com/Rufus31415/Simple-WebXR-Unity/blob/a9060413037d6b5dd985abba29ed17fcd51ff79d/com.rufus31415.simplewebxr/Runtime/Plugins/WebGL/SimpleWebXR.jslib#L129

At next frame, C# code set ```Selected``` property : https://github.com/Rufus31415/Simple-WebXR-Unity/blob/a9060413037d6b5dd985abba29ed17fcd51ff79d/com.rufus31415.simplewebxr/Runtime/Scripts/SimpleWebXR.cs#L712

--> **But** maybe here is the issue : your smartphone only raises event ```Select``` and not ```SelectStart``` so selected is never set to true.

In MRTK, this ```Selected``` property is used : https://github.com/Rufus31415/Simple-WebXR-Unity/blob/a9060413037d6b5dd985abba29ed17fcd51ff79d/SimpleWebXR-Demo/Assets/SimpleWebXR/Scripts/MRTK/SimpleWebXRController.cs#L70

**Comment by gisornator** [comment]
Hmm... weird. Actually, I cannot query for "Selected" because this will always be false. However, I can listen to SelectStart and SelectEnd. So your post makes sense. But for some reason I see Select events when remote debugging on chrome:
![grafik](https://user-images.githubusercontent.com/17878740/151506415-05a21e4b-fa87-4111-94aa-e0b25c2264a7.png)

**Comment by gisornator** [comment, marked as answer]
Okay, classic PICNIC issue here. So I probably didn't understand it correctly in the first place but all UI elements have to be world-canvases and they have to be configured for MRTK, otherwise the other elements like buttons etc. aren't interactable.  The buttons themselves, however, can be standard Unity UI Buttons - as long as the canvas is configured for MRTK! :)

Anyway, I appreciate your help and I love this repo so far. It makes designing AR apps for WebGL much more convenient as we are not bound to certain 3D object types but can also use .FBX etc.

**Comment by Rufus31415** [comment]
Thanks for your kind comment ;) Do not hesitate to contribute to SimpleWebXR if you have time and good ideas !
